src/material/datepicker/calendar.ts

```diff
@@ -149,8 +162,21 @@ export class MatCalendarHeader<D> {
       return this._dateAdapter.getYear(date1) == this._dateAdapter.getYear(date2);
     }
     // Otherwise we are in 'multi-year' view.
-    return Math.floor(this._dateAdapter.getYear(date1) / yearsPerPage) ==
-        Math.floor(this._dateAdapter.getYear(date2) / yearsPerPage);
+    let minYear = 0;
```

Comment:
I would tweak this slightly:
```ts
// We want a range years such that we maximize the number of
// enabled dates visible at once. This prevents issues where the minimum year
// is the last year of a page or the maximum year is the first item of a page.


// We pick a "starting" year such that either the maximum year would be at the end
// or the minimum year would be at the begining of a page.
let startingYear = 0;
if (this.calendar.maxDate) {
  let maxYear = this._dateAdapter.getYear(this.calendar.maxDate);
  startingYear = maxYear - yearsPerPage + 1;
} else if (this.calendar.minDate) {
  startingYear = this._dateAdapter.getYear(this.calendar.minDate);
}

// When the multi-year view is first opened, the active year will be in view.
// So we compute how many years are between the active year and the *slot* where our
// "starting" year year will render when paged into view.
const activeYear = this._dateAdapter.getYear(this.calendar.activeDate);
const activeOffset = this._mod((activeYear - startingYear), yearsPerPage);

// The offset from the active year to the "slot" for the starting year is the
// *actual* first rendered year in the multi-year view, and the lasy year is
// just one yearsPerPage away.
const minYearOfPage = activeYear - activeOffset;
const maxYearOfPage = minYearOfPage + yearsPerPage - 1;

const firstYearInView = this._dateAdapter.getYearName(
    this._dateAdapter.createDate(minYearOfPage, 0, 1));
const lastYearInView = this._dateAdapter.getYearName(
    this._dateAdapter.createDate(maxYearOfPage, 0, 1));
return `${firstYearInView} \u2013 ${lastYearInView}`;
```
* Adds some explanatory comments
* Changes `let` to `const` where appropriate
* Combines two separate `if` blocks into an `if-elseif`
* Changes `minDate` to `startingDate` since it's not necessarily the `minDate`